infrastructure/comms/src/peer_manager/node_id.rs

```diff
@@ -0,0 +1,252 @@
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use derive_error::Error;
+use digest::Digest;
+use tari_crypto::keys::PublicKey;
+
+#[derive(Debug, Error)]
+pub enum NodeIDError {
+    OutOfBounds,
+}
+
+/// Calculate the distance between two node ids using the XOR metric
+pub fn xor_metric(x: &NodeID, y: &NodeID) -> NodeDistance {
+    let mut nd = NodeDistance::new();
+    for i in 0..nd.0.len() {
+        nd.0[i] = x.0[i] ^ y.0[i];
+    }
+    (nd)
+}
+
+#[derive(Clone, Debug, Eq, PartialOrd, Ord)]
+pub struct NodeDistance([u8; 32]);
+
+impl NodeDistance {
+    /// Construct a new zero distance
+    pub fn new() -> NodeDistance {
+        (NodeDistance::from(&[0; 32]))
+    }
+
+    /// Construct a node distance from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeDistance {
+        (NodeDistance(*array))
+    }
+}
+
+impl PartialEq for NodeDistance {
+    fn eq(&self, nd: &NodeDistance) -> bool {
+        for i in 0..self.0.len() {
+            if self.0[i] != nd.0[i] {
+                return false;
+            }
+        }
+        (true)
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct NodeID([u8; 32]);
+
+impl NodeID {
+    /// Construct a new node id on the origin
+    pub fn new() -> NodeID {
+        (NodeID::from(&[0; 32]))
+    }
+
+    /// Construct a node id from an array of 32 bytes
+    pub fn from(array: &[u8; 32]) -> NodeID {
+        (NodeID(*array))
+    }
+
+    /// Derive a node id from a public key: node_id=hash(public_key)
+    pub fn from_public_key<K: PublicKey, D: Digest>(public_key: &K) -> NodeID {
+        let mut array = [0; 32];
+        array.copy_from_slice(D::digest(&public_key.as_bytes()).as_slice());
+        (NodeID(array))
+    }
+
+    /// Generate a node id from a base layer registration using the block hash and public key
+    // pub fn from_baselayer_registration<D: Digest>(......) -> NodeID {
+    // TODO: NodeId=hash(blockhash(with block_height),public key?)
+    // }
+
+    /// Calculate the distance between the current node id and the provided node id using the XOR metric
+    pub fn distance(&self, node_id: &NodeID) -> NodeDistance {
+        (xor_metric(&self, &node_id))
+    }
+
+    /// Find and return the indices of the K nearest neighbours from the provided node id list
+    pub fn closest_indices(&self, node_ids: &Vec<NodeID>, k: usize) -> Result<Vec<usize>, NodeIDError> {
```

Comment:
I know I agreed that this should be in here, but now that I see it I think that the `NodeId` should just be able to give the distance between it and another `NodeId`. This code and `closest` should rather be the routing table's responsibility.

We may not even need this because we can possibly just order the routing table (inserting new `NodeId`s in the correct order) by distance from the node's `NodeId` and so simply pick the first `k` NodeIds from the list.

Reply:
If you agree with the above and it's easier, I think it's ok to leave it here and then move and adjust the code when the routing table is implemented

Reply:
I think this function might have to live here as it might be used by more than just the routing table, such as with the mechanism used by a node when it wants to determine if it should store a data message on behalf of a nearby node. I'm not attached to having it here so we can move it later if needed.